Support RocksDB backend

hygoni opened the pull request:
Hi, this is a RFC for the RocksDB backend implementation.
It passes all existing testcases with RocksDB, except for Queryable.
@jhpratt , Could you please review this?

Some concerns:
- Not sure about the best way to implement an in-memory DB.
   Perhaps we could create a tmpfs file using the tempfile crate?
- It doesn't seem great for store() to fail due to issues with unique ID generation.

Close #17

jhpratt commented on src/backend/rocksdb.rs:
```diff
@@ -0,0 +1,363 @@
+
+    fn connect(&self) -> Result<Self::Connection, Self::Error> {
+        match &self.location() {
+            Location::InMemory => todo!(),
```
```suggestion
            Location::InMemory => unimplemented!(),
```

Let's leave it at this until there's a better way to handle the inability to open a database in memory.

jhpratt commented on src/backend/rocksdb.rs:
```diff
@@ -0,0 +1,363 @@
+    type DeleteStream = DynStream<Result<kv::DeleteResponse, Status>>;
+
+    #[cfg_attr(feature = "tracing", tracing::instrument)]
+    fn initialize(&self, _connection: &Self::Connection) -> Result<(), Self::Error> {
```
If no initialization is necessary, this function and `connect_kv` can be omitted (and the `initialized` field removed. Likewise for the blob backend.

jhpratt reviewed (commented):
It should be noted in comments that a partial success remains a possibility due to transactions not being used.

Additionally, running `cargo fmt` and `cargo clippy` will fix some stylistic issues that are already included via configuration. If any clippy issues are raised from inside the `target` folder, feel free to ignore those.

jhpratt commented:
Regarding your concerns

> Not sure about the best way to implement an in-memory DB. Perhaps we could create a tmpfs file using the tempfile crate?

We can leave this to a follow-up PR.

> It doesn't seem great for store() to fail due to issues with unique ID generation.

The odds of this ever happening is astronomically low, to the point that it will likely _never_ occur. If we fail to generate a unique 64 bit number twice in a row, something is seriously wrong.

jhpratt commented:
Also, to add this to the CLI, you'll need to touch `cli.rs` and `main.rs`. Those changes should be relatively straightforward.

hygoni commented on Cargo.toml:
```diff
@@ -33,16 +33,18 @@ name = "kv"
 ]
 duckdb = ["dep:duckdb"]
 sqlite = ["dep:rusqlite"]
+rocksdb = ["dep:rocksdb", "dep:rand"]
```
I thought replacing every #[cfg(feature = "rocksdb")] with #[cfg(feature = "vendored-rocksdb")] doesn't seem nice? Or if you don't mind, we can go that route.

jhpratt replied:
In `build.rs`, you can add the following

```rust
if cfg!(feature = "vendored-rocksdb") {
    println!("cargo::rustc-cfg=feature=\"rocksdb\"");
}
```

This will mean that you can use `feature = "rocksdb"` within the crate without actually having a `rocksdb` feature.

hygoni commented on src/interop.rs:
```diff
@@ -154,6 +154,31 @@ impl IntoTonicStatus for rusqlite::Error {
+            rocksdb::ErrorKind::Corruption => Status::data_loss(message),
+            rocksdb::ErrorKind::NotSupported => Status::unimplemented(message),
+            rocksdb::ErrorKind::InvalidArgument => Status::invalid_argument(message),
+            rocksdb::ErrorKind::IOError => Status::failed_precondition(message),
```
Oh, it slipped my mind—will adjust.

hygoni commented:
> It should be noted in comments that a partial success remains a possibility due to transactions not being used.

Alright, IIUC using transactions in RocksDB is on hold because the new rocksdb version that includes this pull request (https://github.com/rust-rocksdb/rust-rocksdb/pull/868) hasn't been released on crates.io yet, right?

I noticed there's another crate, [rust-rocksdb](https://crates.io/crates/rust-rocksdb) (maintained by a Cloudflare employee), that stays more up to date and seems to have had the fix since version 0.23 (with the latest being v0.28). Would you prefer to wait for the [rocksdb](https://crates.io/crates/rocksdb) crate to be updated, or should we switch to the rust-rocksdb crate instead?

hygoni commented on src/interop.rs:
```diff
@@ -154,6 +154,31 @@ impl IntoTonicStatus for rusqlite::Error {
+            rocksdb::ErrorKind::Expired => Status::deadline_exceeded(message),
+            rocksdb::ErrorKind::TryAgain => Status::unavailable(message),
+            rocksdb::ErrorKind::CompactionTooLarge => Status::failed_precondition(message),
+            rocksdb::ErrorKind::ColumnFamilyDropped => Status::failed_precondition(message),
```
Sounds reasonable, will adjust. thanks!

jhpratt commented:
Feel free to use `rust-rocksdb`.

hygoni commented:
Thanks for the feedback! I’ve adjusted your suggestions, made some changes to BlobBackend to use transactions for updates, and switched over to rust-rocksdb.

jhpratt commented:
> By the way, I'm a bit confused about what "Partial" support actually means in the README.md.

Basically that bugs should be expected. That's why for DuckDB I included a link to an upstream issue that is a blocker to having full support.

I pushed one additional commit that includes a few miscellaneous cleanups and additions that were made elsewhere in the crate in the meantime. With that, it's ready to merge :slightly_smiling_face:
